File: 12_Advanced_Graphs/05_Alien_Dictionary/0269-alien-dictionary.py
"""
Problem: LeetCode 269 - Alien Dictionary

Key Idea:
The problem can be solved using topological sorting. The given words can be thought of as directed edges between characters of adjacent words. We can build a graph where each character is a node, and the edges represent the order between characters. Then, we can perform topological sorting to find the correct order of characters.

Time Complexity:
- Building the graph takes O(N + E) time, where N is the number of words and E is the number of edges (character comparisons).
- Performing topological sorting takes O(V + E) time, where V is the number of characters and E is the number of edges.
- In the worst case, each character is compared with the next character in all words.
- Overall, the time complexity is O(N + E + V).

Space Complexity:
- The space required for the graph representation is O(V + E), where V is the number of characters and E is the number of edges.
- Other data structures used take constant space.
- Therefore, the space complexity is O(V + E).
"""

from collections import defaultdict, deque


class Solution:
    def alienOrder(self, words: List[str]) -> str:
        graph = defaultdict(list)
        in_degree = defaultdict(int)

        for i in range(len(words) - 1):
            word1, word2 = words[i], words[i + 1]
            for j in range(min(len(word1), len(word2))):
                if word1[j] != word2[j]:
                    graph[word1[j]].append(word2[j])
                    in_degree[word2[j]] += 1
                    break

        queue = deque(char for char, indeg in in_degree.items() if indeg == 0)
        result = []

        while queue:
            char = queue.popleft()
            result.append(char)
            for neighbor in graph[char]:
                in_degree[neighbor] -= 1
                if in_degree[neighbor] == 0:
                    queue.append(neighbor)

        if len(result) < len(in_degree):
            return ""
        return "".join(result)

File: 12_Advanced_Graphs/05_Alien_Dictionary/0269-alien-dictionary.rs
/*
Problem: LeetCode 269 - Alien Dictionary

Key Idea:
The key idea is to model the problem as a directed graph and perform a topological sort to find the order of characters in the alien language.

Approach:
1. Create a directed graph where each character is a node, and there is a directed edge from character 'a' to 'b' if 'a' comes before 'b' in the alien language.
2. To build the graph, compare adjacent words and add directed edges accordingly.
3. Perform a topological sort on the graph to find the order of characters.
   - Initialize an empty result string and a queue for topological sorting.
   - Add nodes with no incoming edges (in-degree = 0) to the queue.
   - While the queue is not empty:
     - Pop a character from the queue, append it to the result string.
     - For each neighbor of the character, decrement its in-degree by 1.
     - If a neighbor's in-degree becomes 0, add it to the queue.
4. If the result string's length is equal to the number of distinct characters, return the result as the correct character order; otherwise, return an empty string, indicating no valid ordering exists.

Time Complexity:
O(N + E), where N is the total number of characters (distinct characters in words) and E is the total number of edges (comparisons between adjacent words).

Space Complexity:
O(N + E) for the graph representation.
*/

// Since this is a hard question and I don't have access to it, can't write error free code